tests/integ_tests/conftest.py

```diff
@@ -14,21 +18,57 @@ def pytest_configure(config):
 
 
 def pytest_collection_modifyitems(config, items):
-    skip_largemodel = pytest.mark.skip(reason="need --largemodel option to run")
-    skip_cpu_only = pytest.mark.skip(reason="skip in --largemodel mode when cpu_only is present")
-    skip_multinode = pytest.mark.skip(reason="Skipped because --multinode was used")
-
-    if config.getoption("--largemodel"):
-        # --largemodel given in cli: only run tests that have largemodel marker
-        for item in items:
-            if "largemodel" not in item.keywords:
-                item.add_marker(skip_cpu_only)
+    parts = config.getoption("--split-n")
+    part = config.getoption("--split-part")
+    split_by = config.getoption("--split-by")
+
+    filtered_items = []
+
+    # Step 1: **Pre-filter tests that would be skipped**
+    for item in items:
+        # Skip tests that are cpu_only if --largemodel is set
+        if config.getoption("--largemodel") and (
+                "largemodel" not in item.keywords or "cpu_only" in item.keywords or "skip" in item.keywords
+        ):
+            continue # Skip adding this test to filtered_items
+        # Skip tests that are largemodel if --largemodel is not set
+        if not config.getoption("--largemodel") and "largemodel" in item.keywords:
+            continue # Skip adding this test to filtered_items
+
+        if config.getoption("--multinode") and (
+                "skip_for_multinode" in item.keywords or "skip" in item.keywords
+        ):
+            continue  # Skip adding this test to filtered_items
+
+        filtered_items.append(item)
+
+    # Step 1.1: **Split by classes if split-by==classes**
+    if split_by == "classes":
+        class_to_tests = {}
+        for item in filtered_items:
+            class_name = item.parent.name
+            if class_name not in class_to_tests:
+                class_to_tests[class_name] = []
+            class_to_tests[class_name].append(item)
+        sorted_classes = sorted(class_to_tests.keys())
+
+    # Step 2: **Determine size of a partition**
+    if split_by == "classes":
+        chunk_size = max(1, len(sorted_classes) // parts)
     else:
-        for item in items:
-            if "largemodel" in item.keywords:
-                item.add_marker(skip_largemodel)
-
-    if config.getoption("--multinode"):
-        for item in items:
-            if "skip_for_multinode" in item.keywords:
-                item.add_marker(skip_multinode)
\ No newline at end of file
+        chunk_size = max(1, len(filtered_items) // parts)
+    print(f"Total tests: {len(filtered_items)}, Splitting into {parts} parts, each with {chunk_size} {split_by}")
+
+    # Step 3: **Determine the range of tests/classes to keep in this partition**
+    start_idx = part * chunk_size
+    end_idx = start_idx + chunk_size
+
+    if part + 1 == parts:
+        # Include all remaining classes/tests in the last partition
+        end_idx = len(filtered_items) if not split_by == "classes" else len(sorted_classes)
```

Comment:
A bit confusing to keep evaluating `split_by == "classes"` inline. Maybe just put all logic in blocks:
if split_by == "classes":
    # Do all logic for class split
elif split_by == "another_option":
    # Do something else
else:
    # Default logic